### ISSUE TITLE
Add ExecuTorch support for RF-DETR

### ISSUE THREAD

**Issue by Borda** [roboflow/rf-detr#1066]
## **Description**
We would like to add support for **ExecuTorch** to enable efficient, on-device inference for the **RF-DETR** model. Bringing ExecuTorch capabilities to RF-DETR will allow deployment on edge devices (mobile, embedded systems) with optimized performance and reduced memory footprints. 

## **Motivation**
As on-device AI deployment becomes crucial for real-time applications, exporting RF-DETR models to the ExecuTorch ecosystem will bridge the gap between high-accuracy object detection and edge hardware constraints. 

This implementation can leverage or align with ongoing ExecuTorch integration efforts within the ecosystem, such as the discussions and implementations highlighted in [react-native-executorch issue #797](https://github.com/software-mansion/react-native-executorch/issues/797).

## **Proposed Solution**
* **Export Pipeline:** Implement an export script/pipeline to convert the RF-DETR model into the ExecuTorch ahead-of-time (AOT) compiler format (`.pte`).
* **Operator Coverage:** Verify that all specific operators used in RF-DETR (e.g., custom attention mechanisms or specialized bounding box heads) are covered by ExecuTorch's native operator library or define custom operators if needed.
* **Validation & Benchmarking:** Provide a basic utility script to validate the numerical accuracy of the exported `.pte` model against the original PyTorch model.

## **Additional Context**
* **Reference Issue:** https://github.com/software-mansion/react-native-executorch/issues/797
* If there are specific hardware backends (e.g., CoreML, QNN, XNNPACK) that should be prioritized for testing the RF-DETR deployment, please let us know in the comments.

**Comment by Borda** [comment]
point of refence: https://github.com/chmjkb/rfdetr-executorch-export

**Comment by anatoly-ryabchenko** [comment]
@Borda can I get assigned to it? 👀

**Comment by anatoly-ryabchenko** [comment]
@Borda 
What backends do we want to support? 
I so far tested on xnnpack (tested on my mac), coreml (tested on my mac), and qnn (tested on qualcomm device cloud). I struggle to make Vulkan work and I'm not sure it's worth it at that point, too many issues that should be fixed upstream.
- xnnpack: has very reasonable logits diff <5e-5
- coreml: has higher diffs, needed to remove 6-rank tensors for coreml; higher diffs are somewhat expected and are aligned with my experience - other models (Ultralytics' YOLO or Insightface models, for example) also have the exact same problem, where outputs are slightly different on CoreML. However, the detections usually have quite similar RP curves.
- qnn: has even higher diffs, also needed to have some last layer to be moved to CPU. One caveat with qnn is that it requires ExecuTorch to be built with qnn enabled, which is not what is shipped by default. So for proper CI/CD testing we will have to build it ourselves and cache it as an artifact to not rebuild it every time.

<img width="1920" height="640" alt="Image" src="https://github.com/user-attachments/assets/d5cb0c95-d4e6-416a-80d6-64cf1439e442" />

As for Vulkan, it's more complicated... besides model surgery to be done (deformable attention multi-output split & batchnorm fusion both crash the partitioner), many operations, most importantly floor, are not supported so during runtime it will have to switch bw CPU & GPU all the time, so I am skeptical that it will work significantly faster than xnnpack on those devices. But even if I do that, when I try to run the .pte it crashes at graph encode stage due to expecting `TensorRef` (fixed tensor) but getting `TENSOR` (runtime computed tensor). These seem to be upstream issues, and I see no value of adding Vulkan as backend at this point honestly.

For now, I still need to do some cleanup and retest, and then I will open a PR with xnnpack / coreml / qnn for now, give instructions about coreml (check your detections) and qnn (check your detections + you need an executorch build with qnn) and will create an issue upstream (in Vulkan).

**Comment by anatoly-ryabchenko** [comment]
Also, do we want a comprehensive COCO test evals to see how much the models are affected by export? If so, which model should I use as a reference - is RF-DETR small enough or I should run it on all of them?

**Comment by anatoly-ryabchenko** [comment]
Okay, PR raised, @Borda please check. 
Besides Vulkan, another shortcoming is failing to do dynamic batch, which also seems to be an upstream (ET) issue. I will create an issue in their repo.
Also, was wondering about additional work in that direction:
1. Direct CoreML export
2. Direct QNN export
3. Direct ET Vulkan export
4. (after fixing ET upstream issue) ET Vulkan 
5. (after fixing ET upstream issue) ET dynamic batch support
6. Dynamic shape (or better to say scale, since aspect ratio should be fixed) support
7. Selecting quant precision
8. Data-driven int8 quant

1/2/7 are quite easy and would be nice to have, 6/8 are also valuable though would take me some more time, while 3-5 probably less valuable but it's still be nice to have them for completeness.
Wdyt? 👀
